language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ModVarToListenerDeclCodeAction.java

```diff
@@ -0,0 +1,137 @@
+            return Collections.emptyList();
+        }
+        Optional<Pair<CaptureBindingPatternNode, String>> captureBindingPatternNodeFilePathPair =
+                findCaptureBindingPattern(matchedNode, context);
+        if (captureBindingPatternNodeFilePathPair.isEmpty() ||
+                captureBindingPatternNodeFilePathPair.get().getLeft().parent().kind()
+                        != SyntaxKind.TYPED_BINDING_PATTERN) {
+            return Collections.emptyList();
+        }
+        TypedBindingPatternNode typedBindingPatternNode =
+                (TypedBindingPatternNode) captureBindingPatternNodeFilePathPair.get().getLeft().parent();
+        List<CodeAction> actions = new ArrayList<>();
+        List<TextEdit> textEdits = new ArrayList<>();
+        Position pos = CommonUtil.toRange(typedBindingPatternNode.lineRange()).getStart();
+        Position insertPos = new Position(pos.getLine(), pos.getCharacter());
+        textEdits.add(new TextEdit(new Range(insertPos, insertPos),
+                SyntaxKind.LISTENER_KEYWORD.stringValue() + " "));
+        String commandTitle = String.format(CommandConstants.CONVERT_MODULE_VAR_TO_LISTENER_DECLARATION,
+                matchedNode.toSourceCode());
+        actions.add(createQuickFixCodeAction(commandTitle, textEdits,
+                captureBindingPatternNodeFilePathPair.get().getRight()));
+        return actions;
+    }
+
+    private Optional<Pair<CaptureBindingPatternNode, String>> findCaptureBindingPattern(Node matchedNode,
+                                                                                        CodeActionContext context) {
+        Optional<Symbol> symbol = context.currentSemanticModel()
+                .flatMap(semanticModel -> semanticModel.symbol(matchedNode));
+        if (symbol.isEmpty() || context.currentSyntaxTree().isEmpty()) {
+            return Optional.empty();
+        }
+        NonTerminalNode foundNode;
+        String uri;
+        if (matchedNode.kind() == SyntaxKind.QUALIFIED_NAME_REFERENCE) {
+            //Todo: we need a proper API to get the syntax tree node. 
+            Optional<Project> project = context.workspace().project(context.filePath());
+            Optional<Location> location = symbol.get().getLocation();
+            if (location.isEmpty() || project.isEmpty() ||
+                    project.get().kind() != ProjectKind.BUILD_PROJECT || symbol.get().getModule().isEmpty()) {
+                return Optional.empty();
+            }
+            Path filePath = project.get().sourceRoot()
+                    .resolve("modules").resolve(symbol.get().getModule().get().id().modulePrefix())
+                    .resolve(location.get().lineRange().filePath());
+            Optional<SyntaxTree> syntaxTree = context.workspace().syntaxTree(filePath);
+            if (syntaxTree.isEmpty()) {
+                return Optional.empty();
+            }
+            foundNode = CommonUtil.findNode(symbol.get(), syntaxTree.get());
```

Comment:
Shall we refactor the `CommonUtil.findNode` API to return an optional? WDYT?

Reply:
That's better. Will do